sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,394 @@
+            # 1D probability distribution along the features
+
+            drop_probs = np.zeros((1, n_features), dtype=np.float64)
+
+            if isinstance(self.missing_distribution, (float, np.floating)):
+                drop_probs[:] = self.missing_distribution / n_features
+            elif isinstance(self.missing_distribution,
+                            (list, tuple, np.ndarray)):
+                missing_distribution = np.asarray(self.missing_distribution)
+                if missing_distribution.shape[0] != n_features:
+                    raise ValueError("The shape of the per feature "
+                                     "drop probabilities vector does not "
+                                     "conform to the number of features, %d"
+                                     % n_features)
+                drop_probs[:] = self.missing_distribution
+            else:
+                raise ValueError("missing_distribution must be a float or "
+                                 " 1D vector (list, tuple or np.ndarray) of "
+                                 "shape (n_features,) or dict of 1D vector / "
+                                 "floats. %r was passed"
+                                 % self.missing_distribution)
+
+        if 1 - drop_probs.ravel().sum() <= np.finfo(float).eps:
+            raise ValueError("The sum of all probabilities in the "
+                             "missing_distribution should sum up to less "
+                             "than 1. The sum was found to be %0.8f"
+                             % drop_probs.ravel().sum())
+
+        drop_counts = (drop_probs * n_values).astype(int)
+
+        if missing_type == 'mcar':
+            self._block_drop_missing_values(X, Ellipsis, 0, drop_counts, rng,
+                                            missing_type, missing_values)
+        else:
+            # For classification, NMAR, consider missing distribution
+            # based on class labels as subsets within data
+            for i, class_i in enumerate(classes):
+                self._block_drop_missing_values(X, y == class_i, i,
+                                                drop_counts, rng,
+                                                missing_type, missing_values)
+        return X
+
+    def _block_drop_missing_values(self, X, samples_mask, encoded_label,
+                                   drop_counts, rng, missing_type,
+                                   missing_values):
+        """Helper to insert missing values in given block (label)"""
+        n_features = X.shape[1]
+        this_block_indices = np.arange(X.shape[0])[samples_mask]
+        for feature in range(n_features):
+            this_n_values = X[samples_mask].shape[0]
+            this_required_n_missing = drop_counts[encoded_label, feature]
+
+            if this_required_n_missing <= 0:
+                continue
+
+            if this_required_n_missing > this_n_values:
```

Comment:
(To make myself clear)

Currently we can say

"I want 40% of missing values to be from class 1".

Or if we give a per feature vector as the key, `{1: [0.1, 0.2, 0]}`

"I want 30% of missing values to be from class 1, where 10% are from feature 1 and the rest 20% are from feature 2".

If we rather have it as piecewise probability of missing it will be

"I want 50% of class 1 samples to go missing."

Or if we give a per feature vector as the key, `{1: [1, 0.5, 0]}`

"I want 100% of class 1 samples to go missing in feature 1, 50% of class 1 samples to go missing in feature 2 and none in feature 3;"